DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,269 @@
+
+### Basic Knowledge
+
+The most important components in Harvester are the custom resources, controllers, webhooks, and the API server.
+
+- Custom Resource (CR): Harvester defines its Kubernetes API using [custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/). The Go API types reside under [pkg/apis](./pkg/apis/). We generate the CustomResourceDefinitions (CRDs) via [go generate](./main.go) and package them under [deploy/charts/harvester-crd](./deploy/charts/harvester-crd). Other Harvester projects may also define their own CRDs.
+- Controller: We use [rancher/wrangler](https://github.com/rancher/wrangler) to implement [custom controllers](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers), which are located in [pkg/controller/master](./pkg/controller/master). Controllers execute the corresponding logic when users create, update, or delete resources.
+- Webhook: Webhooks validate and mutate resources when they are created, updated, or deleted. All rules are located in [pkg/webhook/resources](./pkg/webhook/resources/).
+- API Server: We use [rancher/apiserver](https://github.com/rancher/apiserver?tab=readme-ov-file) to implement the API server, which is located in [pkg/api](./pkg/api) and [pkg/server](./pkg/server). The API server exposes actions used by the Harvester UI.
+
+### Writing Test Cases
+
+When you implement a feature, unit tests are required.
+
+Most Harvester features follow the [Kubernetes controller pattern](https://kubernetes.io/docs/concepts/architecture/controller/), so it's important to test them. Please take a look at the following example to understand basic usage:
+- [support bundle controller](./pkg/controller/master/supportbundle/controller_test.go)
+
+In addition to unit tests, write automated tests where applicable. If the feature requires an automated test, refer to the [Automation Testing](#automation-testing) section for how to write and run them.
+
+## Debugging and Troubleshooting
+
+In general, we need cluster information for debugging. We usually obtain a [support bundle](https://docs.harvesterhci.io/v1.6/troubleshooting/harvester/#generate-a-support-bundle) from users. It contains a lot of useful cluster information.
+
+Then, we use [rancher/support-bundle-kit](https://github.com/rancher/support-bundle-kit) to analyze the support bundle. It's a toolkit for generating and analyzing support bundles for Kubernetes and Kubernetes-native applications.
+
+After building the binary from the project, you can run `./bin/support-bundle-kit-{amd64|arm64} simulator --bundle-path ./supportbundle_xxxx`. It takes some time to build the simulated cluster. Then you can use the kubeconfig from `~/.sim/admin.kubeconfig` to operate the cluster. Since it reconstructs the cluster only from YAML files and logs, commands inside pods will not work.
+
+Finally, refer to the following resources to search for possible solutions.
```

Comment:
Maybe also add a link to the `#harvester-dev` in the Rancher slack, if they need assistance from the maintainers.